request-manager/graph/grapher.go

```diff
@@ -61,58 +61,103 @@ func (gr *Grapher) CheckSequences() (seqGraphs map[string]*Graph, seqErrors map[
+	subsequences := map[string][]string{}
 	for seqName, _ := range seqGraphs {
-		// seqName guaranteed to be in sequenceSpecs; if not, we never
-		// would've built that graph
 		seqSpec := gr.sequenceSpecs[seqName]
+		subseqMap := map[string]bool{}
 
-		missingSets := map[string][]string{} // node name -> list of missing `sets` args
-
-		for nodeName, nodeSpec := range seqSpec.Nodes {
+		for _, nodeSpec := range seqSpec.Nodes {
 			if nodeSpec.IsJob() {
 				continue
 			}
+			subseqs := getNodeSubsequences(nodeSpec, gr.sequenceSpecs)
+			for _, s := range subseqs {
+				subseqMap[s] = true
+			}
+		}
+		for s, _ := range subseqMap {
+			subsequences[seqName] = append(subsequences[seqName], s)
+		}
+	}
 
-			// Check that subsequences actually built
-			// If any subsequence failed, we don't know what
-			// args it sets, so we mark this sequence as having
-			// failed as well.
-			subsequences := getAllSubsequences(nodeSpec, gr.sequenceSpecs)
-			failed := getFailedSubsequences(subsequences, seqGraphs)
-			if len(failed) != 0 {
-				multiple := ""
-				if len(failed) > 1 {
-					multiple = "s"
+	// Keep checking sequences whose dependencies are satisfied until we can't
+	// anymore, either because we've checked them all, or because impossible
+	// dependencies exist.
+	seqsChecked := map[string]bool{}
+	seqsToCheck := map[string]*spec.Sequence{}
+	for k, v := range gr.sequenceSpecs {
+		seqsToCheck[k] = v
+	}
+
+	for len(seqsToCheck) != 0 {
+		newSeqChecked := false
+
+		for seqName, seqSpec := range seqsToCheck {
+			if haveAllDeps(seqsChecked, subsequences[seqName]) {
+				newSeqChecked = true
+				delete(seqsToCheck, seqName)
+				seqsChecked[seqName] = true
+
+				// Only perform sets check if this sequence built
+				// and all its subsequences have passed all checks.
+				if _, ok := seqErrors[seqName]; ok {
+					continue
+				}
+				failed := getFailedSubsequences(subsequences[seqName], seqErrors)
+				if len(failed) != 0 {
+					multiple := ""
+					if len(failed) > 1 {
+						multiple = "s"
+					}
+					seqErrors[seqName] = fmt.Errorf("subsequence%s failed checks: %s", multiple, strings.Join(failed, ", "))
+					continue
 				}
-				seqErrors[seqName] = fmt.Errorf("subsequence%s failed to build: %s", multiple, strings.Join(failed, ", "))
-				delete(seqGraphs, seqName)
-				continue
-			}
 
-			// Compare declared `sets` with what the node
-			// actually sets
-			sets := getDeclaredSets(subsequences, seqSets)
-			missing := getMissingSets(sets, nodeSpec.Sets)
-			if len(missing) != 0 {
-				missingSets[nodeName] = missing
+				// Do sets check.
+				missingSets := map[string][]string{} // node name -> list of missing `sets` args
+				for nodeName, nodeSpec := range seqSpec.Nodes {
+					if nodeSpec.IsJob() {
+						continue
+					}
+					// Compare declared `sets` with what the node
+					// actually sets
+					subseqs := getNodeSubsequences(nodeSpec, gr.sequenceSpecs)
```

Comment:
This is pretty similar to what's happening around line 78. Is there a benefit to doing the 'sets' check here, or would it be possible to move it up there, so we don't have to repeat looping over all the nodes and finding subsequences?

Reply:
If we do it at line 78, the errors from sets checks can be somewhat inconsistent because the loop at 74 doesn't go through sequences in any particular order.

Suppose you have sequences A, B, and C s.t. A calls B calls C (C -> B -> A, since A depends on B depends on C). Furthermore, the node in B that calls C "sets" an arg that no node in C sets, i.e. there's a sets error in B. There are no other errors.

Suppose we do the sets check in the loop at line 74. If the loop at 74 goes through the sequences as A, B, C, then A will pass because `seqErrors` doesn't contain an entry for B. On the next iteration, we fail B, because it contains a sets error. However, if the loop at 74 goes through the sequences as with B before A, then we'll fail B for the sets error, then fail A, because B failed.

If we keep the sets check here (line 127 ig?) then we'll always check dependencies first, and propagate the error across sequences.

Reply:
Ah, okay, let's keep it here then.